pallets/did/src/lib.rs

```diff
@@ -1140,9 +1152,139 @@ pub mod pallet {
+
+			let dispatch_event_payload = result.map(|_| ()).map_err(|e| e.error);
+
+			Self::deposit_event(Event::DidCallDispatched(did_identifier, dispatch_event_payload));
+
+			result
+		}
+
+		/// Store a new DID on chain.
+		///
+		/// The DID identifier is derived from the account ID that submits this
+		/// call. The authentication key must correspond to the account ID that
+		/// submitted this call. For accounts that use the ed25519 and sr25519
+		/// schema, the authentication key must be of the
+		/// `DidVerificationKey::Ed25519` or `DidVerificationKey::Sr25519`
+		/// variant and contains the public key. For Ecdsa accounts, the
+		/// `DidVerificationKey::Ecdsa` variant is calculated by hashing the
+		/// Ecdsa public key.
+		///
+		/// If this call is dispatched by an account id that doesn't correspond
+		/// to a public private key pair, the `DidVerificationKey::Account`
+		/// variant shall be used (Multisig, Pure Proxy, Governance origins).
+		/// The resulting DID can NOT be used for signing data and is therefore
+		/// limited to onchain activities.
+		///
+		/// There must be no DID information stored on chain under the same DID
+		/// identifier. This call will fail if there exists a DID with the same
+		/// identifier or if a DID with the same identifier existed and was
+		/// deleted.
+		///
+		/// The origin for this account must be funded and provide the required
+		/// deposit and fee.
+		///
+		/// Emits `DidCreated`.
+		#[pallet::call_index(16)]
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::create_from_account())]
+		pub fn create_from_account(
+			origin: OriginFor<T>,
+			authentication_key: DidVerificationKey<AccountIdOf<T>>,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			let did_identifier: DidIdentifierOf<T> = sender.clone().into();
+			let auth_key_id: AccountIdOf<T> = authentication_key.into_account();
+			ensure!(auth_key_id == sender, Error::<T>::BadDidOrigin);
+
+			// Make sure that DIDs cannot be created again after they have been deleted.
+			ensure!(
+				!DidBlacklist::<T>::contains_key(&did_identifier),
+				Error::<T>::AlreadyDeleted
+			);
+
+			// There has to be no other DID with the same identifier already saved on chain,
+			// otherwise generate a AlreadyExists error.
+			ensure!(!Did::<T>::contains_key(&did_identifier), Error::<T>::AlreadyExists);
+
+			log::debug!("Creating DID {:?}", &did_identifier);
+
+			let did_entry = DidDetails::from_account(sender.clone(), authentication_key, &did_identifier)
+				.map_err(Error::<T>::from)?;
+
+			Did::<T>::insert(&did_identifier, did_entry.clone());
+
+			let imbalance: CreditOf<T> = <T::Currency as Balanced<AccountIdOf<T>>>::withdraw(
+				&did_entry.deposit.owner,
+				T::Fee::get(),
+				Precision::Exact,
+				Preservation::Preserve,
+				Fortitude::Polite,
+			)?;
+
+			T::FeeCollector::on_unbalanced(imbalance);
+
+			Self::deposit_event(Event::DidCreated(sender, did_identifier));
```

Comment:
Good idea. I also bundled this up with the checks. They where separate since we didn't had transactional extrinsic. Now that they are, we can return errors when ever we want.
[11c4fa4](https://github.com/KILTprotocol/kilt-node/pull/551/commits/11c4fa498c4c457970dda03e512b8968e5f688f0)